setup.py

```diff
@@ -319,6 +318,7 @@ def do_setup():
             'sqlalchemy>=1.1.15, <1.3.0',
             'tabulate>=0.7.5, <=0.8.2',
             'tenacity==4.12.0',
+            'typing',
```

Comment:
-1 - not a fan of needing typing at run time if we can avoid it.

Reply:
We only need this for python <3.5. I can extract `install_requires` into a variable and append this for legacy runtimes.

Reply:
There's a built in syntax for that:

https://github.com/apache/airflow/blob/69e9e9729cb26aaed3c5c4c290d26f3eb40c72fd/setup.py#L294

Can we not use the `if TYPE_CHECKING:` in the code base to not even import the module at run time though?

Reply:
Thanks, updated to install conditional on python version.

I don't see why we couldn't check a flag before importing, but what's the benefit? The `typing` module is trivially fast to import, it's in the standard library for versions of python that aren't soon to be EOL, and for older versions, the wheel is ~25kb.

Reply:
hey, random stranger here -- I found this issue because I was trying to typecheck my own airflow implementation and found that airflow itself doesn't have published stubs, so that caused some issues.  Just wanted to chime in and say that type checking actually can be pretty slow, which is why python 3.7 even introduced a `__future__` import to perform lazy typechecking imports.  See [PEP 563](https://www.python.org/dev/peps/pep-0563) for reference.

What's more, you can't put typechecking imports behind a conditional unless you actually have `typing` installed in the first place to check whether the typechecker is running.  The conventional approach is to just use `'typing;python_version<="3.5"'` as your constraint.

The alternative approach is to do this (I actually do both personally but in theory you can pick one):

```python
def is_type_checking():
    try:
        from typing import TYPE_CHECKING
    except ImportError:
        return False
    return TYPE_CHECKING

MYPY_RUNNING = is_type_checking()
```

Reply:
Thanks @techalchemy. This patch installs `typing` conditional on python version like you described. Conditionally importing expensive modules for type checking makes sense, but as far as I can tell, the only imports I'm doing for type checking here are from the `typing` module: `Sequence`, `Optional`, etc. Those aren't expensive to import, and we have to import `typing` anyway to figure out if we're type checking, like you pointed out--so I'm not sure we need conditional imports yet. What do you think?

Reply:
The point of the function-based approach with the safety net of handling the `ImportError` is to provide some flexibility if it's desired -- typechecking is _very_ slow, and by always performing the requisite imports, you basically ensure that annotations are going to be evaluated at runtime no matter what.  This approach allows you a few options:

1. You can set this up as a configuration parameter and toggle it
2. You can make typing an extra that has a builtin failsafe if it's not installed
3. Even if you do only what I pasted above, you ensure that annotations won't be evaluated at runtime if you put all of your typing imports inside an `if MYPY_RUNNING:` block.

The imports themselves aren't slow, it's the consequence of performing them that causes a performance hit.

Reply:
I don't follow--are you saying that importing values from `typing` affects performance at runtime? What are the consequences of importing those values at runtime, when we're not type checking?